# Notism API - Clean Architecture Implementation

## ğŸ“‹ Table of Contents

1. [Architecture Overview](#architecture-overview)
2. [Layer Definitions](#layer-definitions)
3. [Dependency Rules](#dependency-rules)
4. [Implementation Guidelines](#implementation-guidelines)
5. [Folder Structure](#folder-structure)

---

## Architecture Overview

### The Onion Architecture

```
    â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
    â”‚                      API & Infrastructure                           â”‚
    â”‚          Minimal APIs, Endpoints    â”‚    Data Access, Services      â”‚
    â”‚  â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”  â”‚
    â”‚  â”‚                      Application Layer                        â”‚  â”‚
    â”‚  â”‚              Handlers (CQRS), Request/Response                â”‚  â”‚
    â”‚  â”‚  â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”  â”‚  â”‚
    â”‚  â”‚  â”‚                     Domain Layer                        â”‚  â”‚  â”‚
    â”‚  â”‚  â”‚         Aggregates, Entities, Value Objects,            â”‚  â”‚  â”‚
    â”‚  â”‚  â”‚              Business Rules, Interfaces                 â”‚  â”‚  â”‚
    â”‚  â”‚  â”‚  â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”  â”‚  â”‚  â”‚
    â”‚  â”‚  â”‚  â”‚                  Shared Layer                     â”‚  â”‚  â”‚  â”‚
    â”‚  â”‚  â”‚  â”‚         Common Utilities, Extensions,             â”‚  â”‚  â”‚  â”‚
    â”‚  â”‚  â”‚  â”‚           Constants, Result Pattern               â”‚  â”‚  â”‚  â”‚
    â”‚  â”‚  â”‚  â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜  â”‚  â”‚  â”‚
    â”‚  â”‚  â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜  â”‚  â”‚
    â”‚  â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜  â”‚
    â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
```

## Layer Definitions

### ğŸ¯ Domain Layer (Core)

**Purpose**: Contains business logic organized as aggregates that represent cohesive business concepts and maintain consistency boundaries.

#### Responsibilities

- **Aggregate Roots** - Entry points to aggregate boundaries with business behavior
- **Entities** - Objects with identity within aggregates
- **Value Objects** - Immutable objects representing domain concepts
- **Domain Events** - Represent important business occurrences within aggregates
- **Domain Services** - Business logic that doesn't belong to a single aggregate
- **Business Rules** - Invariants and constraints enforced within aggregate boundaries
- **Repository Interfaces** - Contracts for persisting aggregates

#### Methodology

1. **Identify Aggregates First**: Group related entities around a root entity that owns the lifecycle
2. **Protect Invariants**: All business rules are enforced within the aggregate boundary
3. **Use Value Objects for Concepts**: Email, Password, Money - immutable and self-validating
4. **Raise Domain Events**: Communicate important state changes to other parts of the system
5. **Keep Aggregates Small**: Only include what's needed for consistency
6. **Define Repository Interfaces Here**: The domain dictates what persistence operations it needs

#### Guidelines

| Do | Don't |
|---|---|
| Encapsulate business rules in aggregate methods | Expose setters or allow direct state modification |
| Use private constructors with factory methods | Allow invalid aggregate creation |
| Validate in value object constructors | Perform validation in application layer |
| Keep aggregates focused on one concept | Create large aggregates with many entities |
| Reference other aggregates by ID only | Include references to other aggregate roots |

---

### ğŸ”§ Application Layer (Command/Query Handlers)

**Purpose**: Orchestrates domain operations through CQRS handlers that process commands and queries using MediatR, providing a clear separation between read and write operations.

#### Responsibilities

- **Command Handlers** - Process write operations (Create, Update, Delete)
- **Query Handlers** - Process read operations (Get, Search, List)
- **Request/Response Models** - Input and output models for handlers
- **Validation** - Input validation using FluentValidation
- **Business Orchestration** - Coordinate between multiple domain aggregates
- **Exception Handling** - Throw ResultFailureException for business rule violations
- **Domain Event Handling** - React to domain events for cross-aggregate operations

#### Methodology

1. **One Handler per Use Case**: Each business operation gets its own handler
2. **Validate Early**: Use FluentValidation to check inputs before processing
3. **Orchestrate, Don't Implement**: Handlers coordinate domain objects, not contain business logic
4. **Use Result Pattern**: Return `Result<T>` for explicit success/failure handling
5. **Map at Boundaries**: Transform between domain entities and response DTOs
6. **Feature-Based Organization**: Group related handlers in feature folders (Auth/, User/, etc.)

#### Handler Flow Pattern

```
Request â†’ Validator â†’ Handler â†’ Domain Operations â†’ Repository â†’ Response
```

1. Receive typed request via MediatR
2. Validate using FluentValidation pipeline behavior
3. Load aggregates via repository
4. Execute domain methods (business logic lives in domain)
5. Persist changes
6. Map to response DTO and return

#### Guidelines

| Do | Don't |
|---|---|
| Keep handlers thin and focused | Put business logic in handlers |
| Use specifications for queries | Write raw query logic in handlers |
| Throw `ResultFailureException` for business violations | Return error codes or null |
| Map domain entities to DTOs | Return domain entities directly |
| Use `EnumConverter` for stringâ†”enum conversion | Hard-code enum string values |

---

### ğŸŒ API Layer (Minimal APIs)

**Purpose**: Handles HTTP communication using Minimal APIs with feature-based endpoint organization, delegating request processing to MediatR handlers and managing cross-cutting concerns through middleware.

#### Responsibilities

- **Minimal API Endpoints** - Feature-based HTTP endpoints using MediatR
- **Authentication & Authorization** - JWT-based auth with role-based access control
- **Error Handling** - Global exception handling via middleware
- **Request Validation** - Automatic validation through FluentValidation pipeline
- **API Documentation** - OpenAPI/Swagger integration
- **CORS Configuration** - Cross-origin resource sharing setup

#### Methodology

1. **Delegate to MediatR**: Endpoints only translate HTTP to handler requests
2. **Group by Feature**: Use `MapGroup()` for related endpoints
3. **Handle Cross-Cutting Concerns in Middleware**: Auth, logging, error handling
4. **Use Attributes for Authorization**: `[RequireAdmin]` for role-based access
5. **Consistent Response Pattern**: Success returns data, failure returns error message

#### Endpoint Flow Pattern

```
HTTP Request â†’ Middleware â†’ Endpoint â†’ MediatR â†’ Handler â†’ Response
```

1. Request enters through middleware pipeline (auth, logging)
2. Endpoint extracts request data and creates MediatR request
3. `ISender.Send()` dispatches to appropriate handler
4. Result is translated to HTTP response

#### Guidelines

| Do | Don't |
|---|---|
| Keep endpoints thin - just HTTP translation | Put business logic in endpoints |
| Use `ISender` from MediatR | Inject repositories directly |
| Apply authorization attributes | Check roles manually in endpoint code |
| Return appropriate HTTP status codes | Always return 200 |
| Document with OpenAPI attributes | Skip API documentation |

---

### ğŸ—ï¸ Infrastructure Layer (External Concerns)

**Purpose**: Implements interfaces defined by inner layers and provides aggregate persistence, handling aggregate boundaries correctly while managing external dependencies.

#### Responsibilities

- **Aggregate Repositories** - Persist and retrieve complete aggregates
- **External Services** - HTTP clients, message queues, file systems
- **Event Dispatching** - Publish domain events after aggregate persistence
- **Database Context** - Manage aggregate configurations and relationships
- **Caching** - Cache complete aggregates when appropriate
- **Configuration** - Settings, environment variables

#### Methodology

1. **Implement Domain Interfaces**: Repository implementations live here
2. **Load Complete Aggregates**: Include all entities within the aggregate boundary
3. **Dispatch Domain Events After Save**: Ensure events only fire after successful persistence
4. **Use Entity Type Configurations**: Keep DbContext clean with separate configs
5. **External Services as Interfaces**: Define in Application, implement in Infrastructure

#### Repository Pattern

```
Handler â†’ Repository Interface (Domain) â†’ Repository Implementation (Infrastructure) â†’ DbContext
```

1. Handler depends on `IOrderRepository` (defined in Domain)
2. `OrderRepository` implements the interface (in Infrastructure)
3. Repository uses `DbContext` to persist/retrieve aggregates
4. Domain events are collected and dispatched after `SaveChangesAsync()`

#### Guidelines

| Do | Don't |
|---|---|
| Load aggregates with all their entities | Return partial aggregates |
| Use `Include()` for aggregate children | Lazy load across aggregate boundaries |
| Dispatch events after successful save | Fire events before persistence |
| One repository per aggregate root | Create repositories for entities |
| Use specifications for query logic | Duplicate query logic across repositories |

---

### ğŸ”§ Shared Layer (Common Utilities)

**Purpose**: Provides common utilities, extensions, constants, and helper classes that can be used across all layers without creating dependencies between business layers.

#### Responsibilities

- **Extension Methods** - Common extensions for built-in types
- **Constants** - Application-wide constants
- **Utilities** - Helper classes and static methods
- **Common Models** - Generic result types, pagination models
- **Validation Helpers** - Reusable validation logic
- **Enumerations** - Common enums used across layers

#### Methodology

1. **No Business Logic**: Only generic utilities that could be in any project
2. **Self-Contained**: No dependencies on other project layers
3. **Static Where Appropriate**: Extension methods and utilities can be static
4. **Well-Tested**: These are used everywhere, ensure reliability

#### Guidelines

| Do | Don't |
|---|---|
| Put truly cross-cutting concerns here | Put domain-specific logic here |
| Use for Result pattern, Pagination | Put entity definitions here |
| Keep dependencies minimal | Reference Domain or Application |
| Make utilities generic and reusable | Create single-use helpers |

---

## Implementation Guidelines

### Adding a New Feature Checklist

1. **Domain Layer**
   - [ ] Define/update aggregate root with business methods
   - [ ] Create value objects if needed
   - [ ] Add domain events for important state changes
   - [ ] Define repository interface if new aggregate
   - [ ] Add specifications for query patterns

2. **Application Layer**
   - [ ] Create feature folder (e.g., `Order/CreateOrder/`)
   - [ ] Define Request with `IRequest<Result<Response>>`
   - [ ] Create Validator with FluentValidation rules
   - [ ] Implement Handler orchestrating domain operations
   - [ ] Define Response DTO
   - [ ] Add mapping profile if using AutoMapper

3. **Infrastructure Layer**
   - [ ] Implement repository if new aggregate
   - [ ] Add entity configuration for EF Core
   - [ ] Create migration if schema changed
   - [ ] Implement external service if needed

4. **API Layer**
   - [ ] Add endpoint in appropriate feature file
   - [ ] Configure route, HTTP method, authorization
   - [ ] Add OpenAPI documentation attributes

### Error Handling Strategy

| Layer | Error Type | Approach |
|-------|-----------|----------|
| Domain | Business Rule Violation | Throw domain exception or return failure |
| Application | Validation/Business | Throw `ResultFailureException` |
| Infrastructure | External Service | Wrap in domain-meaningful exception |
| API | HTTP | `ResultFailureMiddleware` converts to 400 |

### Testing Strategy

| Layer | Test Focus | Examples |
|-------|-----------|----------|
| Domain | Business rules, invariants | Aggregate creation, state transitions |
| Application | Handler orchestration | Mock repositories, verify flow |
| Infrastructure | Data access | Integration tests with test DB |
| API | HTTP contracts | Integration tests with test server |

---

## Folder Structure

### Notism API - Current Project Structure

```
notism-api/
â”œâ”€â”€ .dockerignore                                   # Docker ignore patterns
â”œâ”€â”€ .editorconfig                                   # Editor configuration
â”œâ”€â”€ .gitignore                                      # Git ignore patterns
â”œâ”€â”€ .vscode/                                        # VS Code settings
â”œâ”€â”€ Directory.Build.props                           # Global build properties
â”œâ”€â”€ Directory.Packages.props                        # Central package management
â”œâ”€â”€ Dockerfile                                      # API container definition
â”œâ”€â”€ Notism.sln                                      # Solution file
â”œâ”€â”€ README.md                                       # Project documentation
â”œâ”€â”€ docker-compose.yml                              # Container orchestration
â”œâ”€â”€ nginx.conf                                      # Reverse proxy configuration
â”‚
â”œâ”€â”€ docs/
â”‚   â””â”€â”€ architecture.md                            # This document
â”‚
â”œâ”€â”€ src/
â”‚   â”œâ”€â”€ Notism.Shared/                             # ğŸ”§ Shared Layer
â”‚   â”‚   â”œâ”€â”€ Attributes/
â”‚   â”‚   â”œâ”€â”€ Constants/
â”‚   â”‚   â”œâ”€â”€ Exceptions/
â”‚   â”‚   â”‚   â””â”€â”€ ResultFailureException.cs
â”‚   â”‚   â”œâ”€â”€ Extensions/
â”‚   â”‚   â”œâ”€â”€ Models/
â”‚   â”‚   â”‚   â”œâ”€â”€ Pagination.cs
â”‚   â”‚   â”‚   â””â”€â”€ Result.cs
â”‚   â”‚   â””â”€â”€ Notism.Shared.csproj
â”‚   â”‚
â”‚   â”œâ”€â”€ Notism.Domain/                             # ğŸ¯ Domain Layer
â”‚   â”‚   â”œâ”€â”€ Common/
â”‚   â”‚   â”‚   â”œâ”€â”€ AggregateRoot.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ DomainEvent.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ Entity.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ ValueObject.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ Enums/
â”‚   â”‚   â”‚   â”œâ”€â”€ Interfaces/
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ IDomainEvent.cs
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ IRepository.cs
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ ISpecification.cs
â”‚   â”‚   â”‚   â”‚   â””â”€â”€ IUnitOfWork.cs
â”‚   â”‚   â”‚   â””â”€â”€ Specifications/
â”‚   â”‚   â”‚
â”‚   â”‚   â”œâ”€â”€ User/                                  # User Aggregate
â”‚   â”‚   â”‚   â”œâ”€â”€ User.cs                            # Aggregate Root
â”‚   â”‚   â”‚   â”œâ”€â”€ IUserRepository.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ PasswordResetToken.cs              # Entity
â”‚   â”‚   â”‚   â”œâ”€â”€ Enums/
â”‚   â”‚   â”‚   â”‚   â””â”€â”€ UserRole.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ Events/
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ PasswordResetCompletedEvent.cs
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ PasswordResetRequestedEvent.cs
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ UserCreatedEvent.cs
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ UserPasswordChangedEvent.cs
â”‚   â”‚   â”‚   â”‚   â””â”€â”€ UserProfileUpdatedEvent.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ Specifications/
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ ActivePasswordResetTokenByUserIdSpecification.cs
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ PasswordResetTokenByTokenSpecification.cs
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ UserByEmailSpecification.cs
â”‚   â”‚   â”‚   â”‚   â””â”€â”€ UserByIdSpecification.cs
â”‚   â”‚   â”‚   â””â”€â”€ ValueObjects/
â”‚   â”‚   â”‚       â”œâ”€â”€ Email.cs
â”‚   â”‚   â”‚       â””â”€â”€ Password.cs
â”‚   â”‚   â”‚
â”‚   â”‚   â”œâ”€â”€ RefreshToken/                          # RefreshToken Aggregate
â”‚   â”‚   â”‚   â”œâ”€â”€ RefreshToken.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ IRefreshTokenRepository.cs
â”‚   â”‚   â”‚   â””â”€â”€ Specifications/
â”‚   â”‚   â”‚       â”œâ”€â”€ RefreshTokenByTokenSpecification.cs
â”‚   â”‚   â”‚       â””â”€â”€ RefreshTokenByUserIdSpecification.cs
â”‚   â”‚   â”‚
â”‚   â”‚   â””â”€â”€ Notism.Domain.csproj
â”‚   â”‚
â”‚   â”œâ”€â”€ Notism.Application/                        # ğŸ”§ Application Layer
â”‚   â”‚   â”œâ”€â”€ Common/
â”‚   â”‚   â”‚   â”œâ”€â”€ Behaviors/
â”‚   â”‚   â”‚   â”‚   â””â”€â”€ ValidationBehavior.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ Constants/
â”‚   â”‚   â”‚   â”œâ”€â”€ Interfaces/
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ IEmailService.cs
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ IPasswordService.cs
â”‚   â”‚   â”‚   â”‚   â””â”€â”€ ITokenService.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ Mappings/
â”‚   â”‚   â”‚   â”‚   â””â”€â”€ MappingProfile.cs
â”‚   â”‚   â”‚   â””â”€â”€ Utilities/
â”‚   â”‚   â”‚       â””â”€â”€ EnumConverter.cs               # Generic enum converter
â”‚   â”‚   â”‚
â”‚   â”‚   â”œâ”€â”€ Auth/                                  # Authentication Features
â”‚   â”‚   â”‚   â”œâ”€â”€ AuthMappingProfile.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ Login/
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ LoginHandler.cs                # CQRS Handler
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ LoginRequest.cs
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ LoginRequestValidator.cs
â”‚   â”‚   â”‚   â”‚   â””â”€â”€ LoginResponse.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ Register/
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ RegisterHandler.cs
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ RegisterRequest.cs
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ RegisterRequestValidator.cs
â”‚   â”‚   â”‚   â”‚   â””â”€â”€ RegisterResponse.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ RefreshToken/
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ RefreshTokenHandler.cs
â”‚   â”‚   â”‚   â”‚   â””â”€â”€ RefreshTokenRequest.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ RequestPasswordReset/
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ RequestPasswordResetHandler.cs
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ RequestPasswordResetRequest.cs
â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ RequestPasswordResetRequestValidator.cs
â”‚   â”‚   â”‚   â”‚   â””â”€â”€ RequestPasswordResetResponse.cs
â”‚   â”‚   â”‚   â””â”€â”€ ResetPassword/
â”‚   â”‚   â”‚       â”œâ”€â”€ ResetPasswordHandler.cs
â”‚   â”‚   â”‚       â”œâ”€â”€ ResetPasswordRequest.cs
â”‚   â”‚   â”‚       â”œâ”€â”€ ResetPasswordRequestValidator.cs
â”‚   â”‚   â”‚       â””â”€â”€ ResetPasswordResponse.cs
â”‚   â”‚   â”‚
â”‚   â”‚   â”œâ”€â”€ User/                                  # User Management Features
â”‚   â”‚   â”‚   â””â”€â”€ UpdateProfile/
â”‚   â”‚   â”‚       â”œâ”€â”€ UpdateUserProfileHandler.cs
â”‚   â”‚   â”‚       â”œâ”€â”€ UpdateUserProfileRequest.cs
â”‚   â”‚   â”‚       â”œâ”€â”€ UpdateUserProfileRequestValidator.cs
â”‚   â”‚   â”‚       â””â”€â”€ UpdateUserProfileResponse.cs
â”‚   â”‚   â”‚
â”‚   â”‚   â”œâ”€â”€ DependencyInjection.cs
â”‚   â”‚   â””â”€â”€ Notism.Application.csproj
â”‚   â”‚
â”‚   â”œâ”€â”€ Notism.Infrastructure/                     # ğŸ—ï¸ Infrastructure Layer
â”‚   â”‚   â”œâ”€â”€ Common/
â”‚   â”‚   â”‚   â”œâ”€â”€ AppDbContext.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ Repository.cs                      # Generic repository
â”‚   â”‚   â”‚   â””â”€â”€ UnitOfWork.cs
â”‚   â”‚   â”‚
â”‚   â”‚   â”œâ”€â”€ Migrations/                            # EF Core migrations
â”‚   â”‚   â”‚   â”œâ”€â”€ 20241025123456_InitialCreate.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ 20241025234567_AddPasswordResetToken.cs
â”‚   â”‚   â”‚   â””â”€â”€ 20241026034538_ConvertUserRoleToString.cs
â”‚   â”‚   â”‚
â”‚   â”‚   â”œâ”€â”€ RefreshTokens/
â”‚   â”‚   â”‚   â””â”€â”€ RefreshTokenRepository.cs
â”‚   â”‚   â”‚
â”‚   â”‚   â”œâ”€â”€ Services/
â”‚   â”‚   â”‚   â”œâ”€â”€ EmailService.cs
â”‚   â”‚   â”‚   â”œâ”€â”€ PasswordService.cs
â”‚   â”‚   â”‚   â””â”€â”€ TokenService.cs                    # JWT generation with role claims
â”‚   â”‚   â”‚
â”‚   â”‚   â”œâ”€â”€ Users/
â”‚   â”‚   â”‚   â””â”€â”€ UserRepository.cs
â”‚   â”‚   â”‚
â”‚   â”‚   â”œâ”€â”€ DependencyInjection.cs
â”‚   â”‚   â””â”€â”€ Notism.Infrastructure.csproj
â”‚   â”‚
â”‚   â””â”€â”€ Notism.Api/                                # ğŸŒ API Layer
â”‚       â”œâ”€â”€ .env.development                       # Development environment variables
â”‚       â”œâ”€â”€ .env.production                        # Production environment variables
â”‚       â”‚
â”‚       â”œâ”€â”€ Attributes/
â”‚       â”‚   â””â”€â”€ RequireAdminAttribute.cs           # Role-based authorization
â”‚       â”‚
â”‚       â”œâ”€â”€ Endpoints/
â”‚       â”‚   â”œâ”€â”€ AuthEndpoints.cs                   # Authentication endpoints
â”‚       â”‚   â””â”€â”€ UserEndpoints.cs                   # User management endpoints
â”‚       â”‚
â”‚       â”œâ”€â”€ Middlewares/
â”‚       â”‚   â””â”€â”€ ResultFailureMiddleware.cs         # Exception handling
â”‚       â”‚
â”‚       â”œâ”€â”€ Models/
â”‚       â”‚   â””â”€â”€ ErrorResponse.cs
â”‚       â”‚
â”‚       â”œâ”€â”€ Properties/
â”‚       â”‚   â””â”€â”€ launchSettings.json
â”‚       â”‚
â”‚       â”œâ”€â”€ appsettings.json
â”‚       â”œâ”€â”€ appsettings.Development.json
â”‚       â”œâ”€â”€ appsettings.Production.json
â”‚       â”œâ”€â”€ DependencyInjection.cs
â”‚       â”œâ”€â”€ Program.cs                             # Minimal API configuration
â”‚       â””â”€â”€ Notism.Api.csproj
â”‚
â””â”€â”€ tests/                                         # ğŸ§ª Test Projects
    â”œâ”€â”€ Notism.Application.Tests/
    â”‚   â”œâ”€â”€ Auth/
    â”‚   â”‚   â”œâ”€â”€ Login/
    â”‚   â”‚   â”œâ”€â”€ Register/
    â”‚   â”‚   â””â”€â”€ RefreshToken/
    â”‚   â”œâ”€â”€ BaseTest.cs
    â”‚   â””â”€â”€ Notism.Application.Tests.csproj
    â”‚
    â””â”€â”€ Notism.Domain.Tests/                       # Comprehensive domain testing (122 tests)
        â”œâ”€â”€ RefreshToken/
        â”‚   â””â”€â”€ RefreshTokenTests.cs
        â”œâ”€â”€ User/
        â”‚   â”œâ”€â”€ UserTests.cs                        # User aggregate tests
        â”‚   â”œâ”€â”€ EmailTests.cs                       # Email value object tests
        â”‚   â”œâ”€â”€ PasswordTests.cs                    # Password value object tests
        â”‚   â””â”€â”€ UserSpecificationTests.cs           # Specification pattern tests
        â””â”€â”€ Notism.Domain.Tests.csproj
```

### Key Project Structure Highlights

#### ğŸ¯ **Domain-Driven Design Organization**
- **Aggregates**: User and RefreshToken with clear boundaries
- **Value Objects**: Email and Password with business validation
- **Domain Events**: Comprehensive event handling for business actions
- **Specifications**: Encapsulated query logic for all data access

#### ğŸ”§ **CQRS Implementation**
- **Handlers**: Separate command and query handlers for each feature
- **Validation**: FluentValidation validators for each request
- **Responses**: Dedicated response models for each operation
- **Feature Folders**: Authentication and User management organized by business capability

#### ğŸ—ï¸ **Infrastructure Patterns**
- **Repository Pattern**: One repository per aggregate root
- **Unit of Work**: Transaction management and change tracking
- **Services**: External concerns (Email, Password hashing, JWT tokens)
- **Migrations**: Version-controlled database schema changes

#### ğŸŒ **API Design**
- **Minimal APIs**: Feature-based endpoint organization
- **Middleware**: Cross-cutting concerns (error handling, logging)
- **Authorization**: Role-based access control with custom attributes
- **Configuration**: Environment-specific settings management

#### ğŸ§ª **Testing Strategy**
- **Domain Tests**: 122+ tests covering business logic and rules
- **Application Tests**: Handler validation and business flow testing
- **Separation**: Clear test organization matching source structure

This structure demonstrates clean architecture principles with proper separation of concerns, dependency inversion, and comprehensive testing coverage.
